### ISSUE TITLE
Use custom dataset for training by GTX16xx , loss is nan,MAP is 0 

### ISSUE THREAD

**Issue by 666ycy** [ultralytics/ultralytics#1655]
### Search before asking

- [ ] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

当我利用我的GTX1650显卡训练模型，数据集也是完全正确的，可是损失函数一行全是nan,MAP一行全是0，请问有什么解决办法。根据仓库内部分问题需要关闭AMP，，我将如下代码右侧修改为False,可是仍然没有效果，是我关闭的方式不对还是此解决方案无效，恳求各位提供解决方案，并附上解决方式图片或者代码，万分感谢
When I used my GTX1650 graphics card to train the model, the data set was also completely correct, but the loss function row was all Nans and the MAP row was all zeros. Could you tell me the solution? According to some problems in the warehouse, AMP needs to be shut down. I modified the right side of the following code to False, but it still has no effect. Is it the wrong way I closed or the solution is invalid?Please provide solutions, and attach the solution picture or code, thank you very much.

self.amp = bool(self.amp)  # as boolean  →   self.amp = False # as boolean

**Comment by glenn-jocher** [comment]
@666ycy in the latest 8.0.58 version you can force disable amp with `yolo train amp=False`. If this doesn't resolve your issues then they are not AMP-related and may be due to underlying CUDA or torch issues.

**Comment by 666ycy** [comment]
@glenn-jocher  Thank you for your answer.Actually,I want to use the whole project to train rather than the library.My cuda version is 11.3 and torch version is 1.12.1.I want to close amp in the project,but I don‘t know how to do

**Comment by glenn-jocher** [comment]
@666ycy just pass amp=False to any training to disable AMP. If you are having issues locally you should try training in our Docker image:

## Environments

YOLOv8 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/GCP-Quickstart)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/AWS-Quickstart)
- **Docker Image**. See [Docker Quickstart Guide](https://github.com/ultralytics/yolov5/wiki/Docker-Quickstart) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

**Comment by glenn-jocher** [comment]
Thank you for the update, @666ycy. It's possible that the issue you're experiencing could be related to the AMP (Automatic Mixed Precision) optimization process. If your loss function is returning NaN values, it can cause the MAP (Mean Average Precision) score to go to zero. Disabling AMP may help to solve this kind of issue. Also, make sure that your dataset is in the correct format and your system requirements meet the minimum specifications for training with YOLOv8. Finally, note that we are constantly working on updates and improvements for YOLOv8, so we encourage you to check for new releases and updates on the GitHub repository.

**Comment by Monday-Leo** [comment]
I face the same problem. I set the amp=False, then the train loss is correct. But in the evaluate process, the Map is still zero. How to solve this problem. My gpu is 1660S with torch 1.13 cuda 11.6. @glenn-jocher 

`      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      1/300      3.82G      1.441      1.605      1.058         34        448: 100%|██████████| 405/405 [05:53<00:00,  1.15it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 25/25 [00:21<00:00,  1.14it/s]
                   all       1590       3229          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      2/300      3.84G       1.42     0.9724      1.043         12        448: 100%|██████████| 405/405 [05:52<00:00,  1.15it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 25/25 [00:21<00:00,  1.14it/s]
                   all       1590       3229          0          0          0          0`

**Comment by Monday-Leo** [comment]
If run the eval command alone it gives a correct result as follows.

Ultralytics YOLOv8.0.82  Python-3.7.0 torch-1.13.0+cu116 CUDA:0 (NVIDIA GeForce GTX 1660 SUPER, 6144MiB)
YOLOv8n summary (fused): 168 layers, 3006038 parameters, 0 gradients, 8.1 GFLOPs
val: Scanning E:\code\YOLOv8-15000\datasets\datasets\labels\val.cache... 1590 images, 107 backgrounds, 0 corrupt: 100%|██████████| 1590/1590 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 100/100 [00:21<00:00,  4.57it/s]
                   all       1590       3229      0.795      0.677      0.749      0.379
                  body       1590       1677      0.756      0.875       0.89      0.513
                  head       1590       1552      0.834      0.479      0.608      0.245

**Comment by tongchangD** [comment]
@Monday-Leo  我遇到了与这相似的情况，但是在验证数据上出现NAN后，然后保存的模型都不可以用了，在测试或其他情况MAP值都是0了，请问你解决这种方案了吗

**Comment by glenn-jocher** [comment]
@tongchangD 欢迎来到 YOLOv8！这个问题可能与 AMP (Automatic Mixed Precision) 优化过程有关。如果您的损失函数返回 NaN 值，会导致 MAP (Mean Average Precision) 得分为零。您已经将 AMP 禁用，但我建议您确认您的数据集是否处于正确的格式，并且您的系统要求符合在 YOLOv8 中训练的最低规格。请尝试在验证阶段运行单独的评估命令，例如`yolo val --task <task> --weights <weights> --data <data> --img <img_size>`，通常可以获得更准确的指标。此外，我们还在不断改进 YOLOv8，因此建议您在 GitHub 存储库中查找新版本和更新。希望这可以帮助您解决问题。

**Comment by tongchangD** [comment]
@glenn-jocher  您好，我试过了`yolo val --task <task> --weights <weights> --data <data> --img <img_size>` 这种方式，val的模型结果依旧是NaN值，伴随着MAp值也是0，可我以如下方式重新保存模型参数，结果就有了
```
import torch
last = torch.load("./weights/last.pt")
best = torch.load("./weights/best.pt")
other = best
other["model"] = last["model"]
torch.save(other,"./best_cun_last.pt")
weights= "./last_new.pt"
```
我将 <weights> 参数替换为 如上的`weights= "./last_new.pt"` ，模型结果就有了，不再是NaN，而且MAp值也正常了，
请问这是和解？，为什么？

祝好
tongchangD

**Comment by glenn-jocher** [comment]
@tongchangD 你好，tongchangD! 感谢您联系我们。根据您的描述，在验证时，您的模型返回 NaN 值，但当您使用代码重写模型参数后，您的模型就能够正常运行了。这可能是由于模型训练期间程序出现了意料之外的停止造成的。您可以考虑使用 last.pt 模型继续训练，这样可能会更好一些。此外，如果您在训练过程中遇到了意外错误，例如程序崩溃或强制终止，这可能会影响您的模型性能。请注意检查您的代码并避免这些不良情况。希望这可以解决您的问题。

**Comment by glenn-jocher** [comment]
@tongchangD 您好！感谢您联系我们。如果您在验证时遇到 NaN 的问题，我建议您先仔细检查数据集是否处于正确的格式，并检查训练期间的输入是否正确。此外，这可能与 AMP (Automatic Mixed Precision) 优化过程有关。如果您的损失函数返回 NaN 值，会导致 MAP (Mean Average Precision) 得分为零。如果您遇到了这个问题，禁用 AMP 可能会解决它。您还可以尝试在训练之前运行数据集分析以检查数据集的质量。对于保存的模型存在错误的问题，建议您在训练期间将模型保存在较短的时间间隔内，并进行多次迭代的模型验证以确保模型的有效性。希望这些建议能够帮助您，祝您成功！

**Comment by ChristopherRogers1991** [comment]
There is an issue about training on Mac M1 GPUs that results in the `Box(P R mAP50 mAP50-95)` values all being 0 as well - I found that first, and left a comment about hitting the same thing on an Nvidia GPU: https://github.com/ultralytics/ultralytics/issues/3962#issuecomment-1660008876.

I was also having issues with AMP causing `nan`s, and was looking looking to see if that had been reported when I found this issue. With `amp=False`, the loss values become valid, and my model seems to work (can correctly detect objects), but the `Box(P R mAP50 mAP50-95)` all remain 0.

**Comment by ChristopherRogers1991** [comment]
Potentially worth noting - I was able to get help in the discord chat (https://discord.com/channels/1089800235347353640/1135559576775835668/1135873639183564930), but it took some digging (had to diff the cpu and gpu outputs, to find the `nan`s). It might be good to cause a hard failure if those loss values are `nan`. Otherwise, it looks like the training is completely successful, but the model just doesn't work, which is fairly difficult to troubleshoot.

**Comment by glenn-jocher** [comment]
@ChristopherRogers1991 hi there! Thanks for sharing this. Your observations could be valuable in helping us improve our YOLOv8 model error handling and training feedback. I understand that dealing with instances of 'nan' loss values can be quite tricky to troubleshoot without the system reporting a clear error.

Indeed, placing mechanisms to cause a hard failure when loss values become 'nan' could be a good idea. It could offer an early indication of some issues with the training process. However, this could also inadvertently stop training processes that might still end up successful despite the occurrence of 'nan' loss values.

I appreciate your suggestion to improve the troubleshooting process. It's valuable feedback for us and something we could consider for future updates. Keep in mind that machine learning models can sometimes behave unexpectedly due to the complexity of the data or the training process and it's always beneficial to have more robust error handling. Thanks for staying involved with the YOLOv8 community and contributing with valuable insights. Your efforts help us make our models better for everyone.

**Comment by bluceliuljx** [comment]
i have the same error
![image](https://github.com/ultralytics/ultralytics/assets/40584477/81a5e4d2-7029-4f85-951d-e8bbb0568afb)

the train process is not the error,
however, 
in the val process , data input the frist conv , x is nan

**Comment by bluceliuljx** [comment]
终于发现了
数据经过conv后，再经过BN层，在BN层后，出现了nan现象
![image](https://github.com/ultralytics/ultralytics/assets/40584477/ebf47c24-e17f-4f59-856a-78dff9b5b0a8)

**Comment by glenn-jocher** [comment]
@bluceliuljx 你好！看来你已经找到了问题可能出在哪里。数据在经过卷积层(Conv)之后，在批归一化(BN)层产生了 NaN。这可能是由于批归一化层处理过程中出现了数学上的不稳定，比如除以接近零的标准差。

对于这种情况，你可能需要先确保你的数据集在输入模型前没有包含任何无效或异常的值。接下来，你可以尝试修改模型的批归一化层配置，可能可以通过调整学习率或在训练早期引入一些正则化来解决这个问题。

请注意，处理这种问题可能需要一些试验和调整。希望这个解答能帮到你，祝你好运！

**Comment by bluceliuljx** [comment]
在 nn.BatchNorm2d 中，设置参数 track_running_stats=False ， 能解决此问题，但是在部署阶段，转到onnx时，合并conv bn时，fuse操作上还需要补充操作的
![Uploading image.png…]()

**Comment by glenn-jocher** [comment]
@bluceliuljx 您好！非常感谢您与我们分享您的解决方案。设置 `track_running_stats=False` 在 `nn.BatchNorm2d` 中确实可以在某些情况下解决这个问题，此参数可以控制批规范化层是否应计算并存储运行中的均值和方差以用于评估模式。

对于您在部署阶段遇到的问题，转换到 ONNX 时，融合卷积和规范化层的操作可能需要些额外的步骤。ONNX 的 fuse 操作可能不能直接处理 `track_running_stats=False`，您可能需要在完成 fuse 操作后手动完成额外的步骤。这可能会涉及到明确地处理 ONNX 中的 BN 层。

一般来说，解决这些问题的最好方法是明确地理解和处理在转换过程中可能遇到的问题，而不仅仅是依赖自动融合等工具提供的功能。这可能需要对 ONNX 和您的模型的内部工作有深入的理解。

希望这个答复能对您有所帮助。如果您有进一步的问题或疑虑，欢迎继续留言讨论。感谢您对 YOLOv8 的关注和贡献！
